control-plane/agents/src/bin/core/volume/scheduling.rs

```diff
@@ -143,3 +171,90 @@ pub(crate) async fn resizeable_replicas(
         .map(|item| item.state().clone())
         .collect()
 }
+
+/// Return true if affinity is requested in the volume spec.
+pub(crate) fn is_affinity_requested(suitable_pools: GetSuitablePools) -> bool {
+    match suitable_pools.spec().topology.as_ref() {
+        Some(topology) => {
+            let node_affinity = match &topology.node {
+                Some(node_topology) => match node_topology {
+                    NodeTopology::Labelled(label) => {
+                        let mut matching_key = false;
+                        for key in label.affinity.keys() {
+                            if label.inclusion.contains_key(key) {
+                                matching_key = true;
+                            }
+                        }
+                        !matching_key && !label.affinity.is_empty()
+                    }
+
+                    NodeTopology::Explicit(_) => false,
+                },
+                None => false,
+            };
+            let pool_affinity = match &topology.pool {
+                Some(pool_topology) => match pool_topology {
+                    PoolTopology::Labelled(label) => {
+                        let mut matching_key = false;
+                        for key in label.affinity.keys() {
+                            if label.inclusion.contains_key(key) {
+                                matching_key = true;
+                            }
+                        }
+                        !matching_key && !label.affinity.is_empty()
+                    }
+                },
+                None => false,
+            };
+            node_affinity || pool_affinity
+        }
+        None => false,
+    }
+}
+
+/// Return a list of pre sorted pools in case of affinity.
+pub(crate) fn second_pass_filter(
+    mut pools: Vec<PoolWrapper>,
+    suitable_pools: GetSuitablePools,
+    registry: &Registry,
+) -> Vec<Vec<PoolWrapper>> {
+    let mut map: HashMap<String, Vec<PoolWrapper>> = HashMap::new();
+    for pool in pools.iter_mut() {
+        let id = pool.state().id.clone();
+        let spec = registry.specs().pool(&id).unwrap_or_default();
+        let pool_labels = spec.labels.clone().unwrap_or_default();
+
+        // Get the affinity labels from the pool topology
+        let pool_affinity = match &suitable_pools.spec().topology.as_ref().unwrap().pool {
+            Some(pool_topology) => match pool_topology {
+                PoolTopology::Labelled(label) => label.affinity.clone(),
+            },
+            None => HashMap::new(),
+        };
+
+        // performance: gold and zone: us-west, then the map key is "performance:gold;zone:us-west;"
```

Comment:
hmm I don't understand this, why is the map key `performance:gold;zone:us-west;`?
Shouldn't there be 2 separate keys?

Reply:
It clubs all the keysand maps and create the key of the hash map inorder to club all pool wrappers having the same key value pair.

Reply:
But why is this required, can't it simply put it as a map?

Reply:
Let's get a conclusion on this question so we can merge on this PR?